File: backend/horse-racing-management/src/main/java/com/horseracing/horseracingmanagement/module/entity/TrainingContract.java
package com.horseracing.horseracingmanagement.module.entity;

import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;

import java.math.BigDecimal;
import java.time.Instant;
import java.time.LocalDate;

@Builder
@AllArgsConstructor
@NoArgsConstructor
@Getter
@Setter
@Entity
@Table(name = "training_contract")
public class TrainingContract {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "id", nullable = false)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "horse_id", nullable = false)
    private Horse horse;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "trainer_id", nullable = false)
    private Trainer trainer;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "owner_id", nullable = false)
    private HorseOwner owner;

    @Column(name = "start_date")
    private LocalDate startDate;

    @Column(name = "end_date")
    private LocalDate endDate;

    @Column(name = "fee", nullable = false)
    private BigDecimal fee;  // tổng phí

    @Column(name = "fee_type", length = 20)
    private String feeType;  // MONTHLY, PERIOD

    @Column(name = "owner_note", columnDefinition = "TEXT")
    private String ownerNote;  // ghi chú từ owner khi gửi

    @Column(name = "trainer_note", columnDefinition = "TEXT")
    private String trainerNote;  // ghi chú từ trainer khi phản hồi

    @Column(name = "status", length = 30)
    private String status;
    // PENDING    → owner gửi, chờ trainer
    // ACCEPTED   → trainer chấp nhận, tiền bị giữ (escrow)
    // REJECTED   → trainer từ chối
    // ACTIVE     → đang trong thời gian train
    // COMPLETED  → hết hạn hợp đồng
    // CANCELLED  → owner hủy (trước khi trainer accept)

    @CreationTimestamp
    @Column(name = "created_at")
    private Instant createdAt;

    @Column(name = "accepted_at")
    private Instant acceptedAt;
}

File: frontend/horse-racing-management-FE/src/types/trainingContract.ts
// Mirrors backend TrainingContract lifecycle. ACCEPTED is declared in the entity but
// never produced — trainer acceptance jumps straight to ACTIVE. COMPLETED is produced
// by the daily auto-complete job once a contract's end date passes.
export type TrainingContractStatus =
  | 'PENDING'
  | 'ACTIVE'
  | 'REJECTED'
  | 'CANCELLED'
  | 'ACCEPTED'
  | 'COMPLETED'
  | string;

export type TrainingFeeType = 'MONTHLY' | 'PERIOD' | string;

export interface TrainingContract {
  id: number;
  horseId: number;
  horseName?: string;
  horseAvatarUrl?: string;
  trainerId: number;
  trainerName?: string;
  trainerAvatarUrl?: string;
  ownerId: number;
  ownerName?: string;
  startDate?: string; // LocalDate — 'yyyy-MM-dd'
  endDate?: string;
  fee: number;        // BigDecimal on the wire — coerce with Number() before formatting
  feeType?: TrainingFeeType;
  ownerNote?: string;
  trainerNote?: string;
  status: TrainingContractStatus;
  createdAt?: string;
  acceptedAt?: string;
  daysRemaining?: number | null; // null unless ACTIVE
  totalDays?: number | null;
  progressPercent?: number | null;
}

export interface SendTrainingContractPayload {
  horseId: number;
  trainerId: number;
  startDate: string; // 'yyyy-MM-dd'
  endDate: string;
  fee: number;
  feeType: TrainingFeeType;
  ownerNote?: string;
}
